devito/operator.py

```diff
@@ -150,15 +157,22 @@ def arguments(self, **kwargs):
         # Derivation. It must happen in the order [tensors -> dimensions -> scalars]
         for i in self.parameters:
             if i.is_TensorArgument:
-                assert(i.verify(kwargs.pop(i.name, None)))
-        runtime_dimensions = [d for d in self.dimensions if d.value is not None]
+                assert(i.verify(kwargs.pop(i.name, None), self.argument_engine))
+        runtime_dimensions = [d for d in self.dimensions if not d.is_Fixed]
         for d in runtime_dimensions:
-            d.verify(kwargs.pop(d.name, None))
+            d.verify(kwargs.pop(d.name, None), self.argument_engine, enforce=True)
         for i in self.parameters:
             if i.is_ScalarArgument:
-                i.verify(kwargs.pop(i.name, None))
-
-        dim_sizes = OrderedDict([(d.name, d.value) for d in runtime_dimensions])
+                i.verify(kwargs.pop(i.name, None), self.argument_engine, enforce=True)
+        dim_sizes = {}
+        for d in runtime_dimensions:
```

Comment:
perhaps this ``extent`` calculation could be a ``Dimension`` property?

Reply:
Dimensions are global objects shared between different Operators. This is an intrinsic property of an Operator.

Reply:
I disagree; that'd be a property of a *specific Operator invocation*, not of the Operator itself. And also, so would be ``.limits``. Finally, let's not forget that a Dimension **is** an ArgProvider, so... We can argue whether that should be the case, but that's another story

Reply:
So if it is a property of the invocation, it belongs right here, no? 
I don't understand what that ArgProvider thing is about but it seems off topic

Reply:
Being an ``ArgProvider``, it seemed kinda natural to have an ``extent`` method.

I'm just concerned that the ``.arguments`` method is becoming more and more complicated, so it would be useful (for many purposes) to "offload" some bit of functionalities to other modules, if/where possible. Mine was just a suggestion in this direction (perhaps a conceptually wrong one, but I'm not that sure).

Reply:
And that's exactly what the ArgumentEngine is slated to do

Reply:
What's the conclusion from this discussion? Are there changes required to this PR or is the increasing complexity of .arguments being highlighted for some TLC (ie should an issue be articulated explaining the problem and we can assign the task to someone)?